pkg/image/registry/imagestreamtag/rest.go

```diff
@@ -40,14 +48,45 @@ func nameAndTag(id string) (name string, tag string, err error) {
 		name = segments[0]
 		tag = segments[1]
 		if len(name) == 0 || len(tag) == 0 {
-			err = errors.NewBadRequest("ImageStreamTags must be retrieved with <name>:<tag>")
+			err = kapierrors.NewBadRequest("ImageStreamTags must be retrieved with <name>:<tag>")
 		}
 	default:
-		err = errors.NewBadRequest("ImageStreamTags must be retrieved with <name>:<tag>")
+		err = kapierrors.NewBadRequest("ImageStreamTags must be retrieved with <name>:<tag>")
 	}
 	return
 }
 
+func (r *REST) List(ctx kapi.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
+	imageStreams, err := r.imageStreamRegistry.ListImageStreams(ctx, labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	matcher := MatchImageStreamTag(label, field)
+
+	list := &api.ImageStreamTagList{}
+	for _, currIS := range imageStreams.Items {
+		for currTag := range currIS.Status.Tags {
+			uncast, err := r.Get(ctx, api.JoinImageStreamTag(currIS.Name, currTag))
```

Comment:
> This is actually 2N+1, which is even worse.
> 
> To perform get, we fetch 2 resources, the IS and the Image. To perform list, we make 2N+1 calls. Even N+1 is insanely inefficient. I'm just not sure we can afford to implement it right now.

As a user, I'd rather have something slow that works than not having it at all and trying to stitch a similar piece together.  There's no way to avoid fetching the image, but its pretty simple to avoid refetching the imagestream.  This definitely highlights the need to fill in labels on an `istag` to allow an efficient query by image stream.

If someone needs to list the istags, this can be made better than what they could do client side and its a lot easier for them.